internal/framework/fetch/fetch.go

```diff
@@ -0,0 +1,569 @@
+	if err == nil {
+		s3Client = s3.NewFromConfig(cfg)
+	}
+
+	httpClient := &http.Client{
+		Timeout: defaultTimeout,
+	}
+
+	return &DefaultFetcher{
+		s3Client:   s3Client,
+		httpClient: httpClient,
+	}, nil
+}
+
+// NewDefaultFetcherWithS3Client creates a new DefaultFetcher with a custom S3 client.
+// This is primarily used for testing with fake S3 clients.
+func NewDefaultFetcherWithS3Client(s3Client S3Client) *DefaultFetcher {
+	httpClient := &http.Client{
+		Timeout: defaultTimeout,
+	}
+
+	return &DefaultFetcher{
+		s3Client:   s3Client,
+		httpClient: httpClient,
+	}
+}
+
+// GetRemoteFile fetches a remote file with retry logic and optional validation.
+// Supports both HTTP(S) and S3 URLs with automatic protocol detection.
+func (f *DefaultFetcher) GetRemoteFile(targetURL string, opts ...Option) ([]byte, error) {
+	ctx := context.Background()
+
+	// Apply options to defaults
+	options := defaults()
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	// Route to appropriate fetcher based on URL scheme
+	if strings.HasPrefix(targetURL, "s3://") {
+		return f.fetchS3File(ctx, targetURL, options)
+	}
+
+	if strings.HasPrefix(targetURL, "http://") || strings.HasPrefix(targetURL, "https://") {
+		return f.fetchHTTPFile(ctx, targetURL, options)
+	}
+
+	return nil, fmt.Errorf("unsupported URL scheme: %s (supported: http://, https://, s3://)", targetURL)
+}
+
+// fetchS3File fetches a file from S3 using the AWS SDK.
+func (f *DefaultFetcher) fetchS3File(ctx context.Context, s3URL string, opts options) ([]byte, error) {
+	if f.s3Client == nil {
+		return nil, fmt.Errorf("S3 client not available - AWS credentials may not be configured")
+	}
+
+	bucket, key, err := parseS3URL(s3URL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid S3 URL %s: %w", s3URL, err)
+	}
+
+	backoff := createBackoffConfig(opts.retryBackoff, opts.retryAttempts, opts.retryMaxDelay)
+	var lastErr error
+	var result []byte
+
+	err = wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		data, err := f.getS3Object(ctx, bucket, key, opts.timeout)
+		if err != nil {
+			lastErr = &S3Error{Bucket: bucket, Key: key, Err: err}
+			// Intentionally return nil error to signal retry mechanism to continue
+			return false, nil //nolint:nilerr // Retry on S3 errors
+		}
+
+		if opts.checksumEnabled {
+			if err := f.validateS3FileContent(ctx, data, bucket, key, opts); err != nil {
+				lastErr = err
+				// Don't retry on checksum mismatches
+				var checksumErr *ChecksumMismatchError
+				if errors.As(err, &checksumErr) {
+					return false, err // Stop retrying
+				}
+				return false, nil // Retry on other checksum errors
+			}
+		}
+
+		result = data
+		return true, nil
+	})
+
+	if result != nil {
+		return result, nil
+	}
+
+	// Return the most meaningful error
+	if lastErr != nil {
+		return nil, lastErr
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch S3 file after retries: %w", err)
+	}
```

Comment:
Looks like the only way this gets returned is if the checksum doesn't match. If so, then this message could probably be tweaked a bit.